Add suppress_ifgroup option in routing policy

bacher09 opened the pull request:
This adds `SuppressIfGroup` option in the `RoutingPolicyRule` section which
has the same semantics as `suppress_ifgroup` in `ip rule` command.

This was previously discussed in #14724 which added `suppress_prefixlen`, but not `supppress_ifgroup` since setting interface groups wasn't implemented back then.

I also wanted to discus one change in a current interface group implementation. Both [linux kernel](https://github.com/torvalds/linux/blob/master/include/linux/netdevice.h#L2012) and [iproute2](https://git.kernel.org/pub/scm/network/iproute2/iproute2.git/tree/ip/iprule.c#n410) is using `int` for a group field, but [systemd](https://github.com/systemd/systemd/blob/main/src/udev/udev-netlink.h#L20) is using `uint` type for this field, which I think is wrong.  Netlink interface uses `uint` as well, but when [kernel receives](https://github.com/torvalds/linux/blob/9e9fb7655ed585da8f468e29221f0ba194a5f613/net/core/fib_rules.c#L611) this value it casts it into `int` and `-1` has a special meaning there. But with current `systemd` implementation `-1` can be accidentally set by specifying `429496729` as group value. 

For example here's what I did:
```
[Match]
Name=test-ifgroup

[Network]
Address=10.0.0.1/24
LLMNR=no
IPv6AcceptRA=no
MulticastDNS=no

[Link]
Group=4294967295
```

and I got this result from `ip link show test-ifgroup`

```
$ ip link show test-ifgroup
4: test-ifgroup: <BROADCAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN mode DEFAULT group -1 qlen 1000
    link/ether 12:34:56:78:9a:bc brd ff:ff:ff:ff:ff:ff
```

So when I'm parsing `SuppressIfGroup` field I'm restricting values in `[0, 2147483647]` range and I do believe that same change should be done for a `Group` as well (and probably kernel shouldn't accept these values too).

keszybz commented on man/systemd.network.xml:
```diff
@@ -1281,6 +1281,12 @@ Table=1234</programlisting></para>
           </listitem>
         </varlistentry>
+        <varlistentry>
+          <term><varname>SuppressIfGroup=</varname></term>
```
I'm not crazy about the option name. Something more self-explanatory would be nice, but I don't have a good idea.

Please explain what this option does here, and what the valid values are.

A good start is the kernel commit message:
   This change adds the ability to suppress a routing decision based upon the
    interface group the selected interface belongs to. This allows it to
    exclude specific devices from a routing decision.

bacher09 replied:
Hey, sorry for this I was going to update it on a previous weekend but didn't had enough time so it was left like this (that's partially why I named it as a `WIP`, to prevent from merging). I published it just because wanted to know opinion if I could change `Group` type from a `uint32_t` to `int32_t`.

keszybz commented on src/network/networkd-routing-policy-rule.c:
```diff
@@ -1622,6 +1643,54 @@ int config_parse_routing_policy_rule_suppress_prefixlen(
+
+        r = safe_atoi32(rvalue, &n->suppress_ifgroup);
+        if (r < 0) {
+                n->suppress_ifgroup = -1;
```
You say "ignoring", but in fact the value is reset. Please drop this line.

keszybz commented on src/network/networkd-routing-policy-rule.c:
```diff
@@ -1622,6 +1643,54 @@ int config_parse_routing_policy_rule_suppress_prefixlen(
+        }
+        if (n->suppress_ifgroup < 0) {
+                log_syntax(unit, LOG_WARNING, filename, line, r,
+                           "Failed to parse SuppressIfGroup=, value should be in range: [0, 2147483647]");
```
Hmm, this message is misleading, since it *was* parsed successfully. Just say "Invalid SuppressIfGroup= value %"PRIi32", must be in the range 0…2147483647". 

Also use a temporary variable and avoid setting the output variable on failure.

keszybz commented on src/network/networkd-routing-policy-rule.h:
```diff
@@ -51,6 +51,7 @@ typedef struct RoutingPolicyRule {
         struct fib_rule_uid_range uid_range;
 
         int suppress_prefixlen;
+        int suppress_ifgroup;
```
Why is this an int? If you make this `int32_t`, you can drop the casts in various places.

keszybz reviewed (changes requested):
Looks reasonable in general.

bacher09 commented:
Please check it again, I've updated it (should address all the comments), also I changed type of link `Group` to `int32_t` for a reasons I highlighted above.

yuwata commented on src/network/networkd-routing-policy-rule.c:
```diff
@@ -1052,6 +1064,15 @@ int manager_rtnl_process_rule(sd_netlink *rtnl, sd_netlink_message *message, Man
+                return 0;
+        }
+        if (r >= 0)
+                tmp->suppress_ifgroup = (int) suppress_ifgroup;
```
Please check the range of the received value, and please use `int32_t` instead of `int`:
```
if (r >= 0) {
        if (suppress_ifgroup > INT32_MAX) {
                log_warning("rtnl: received invalid FRA_SUPPRESS_IFGROUP attribute value, ignoring: %"PRIu32, suppress_ifgroup);
                return 0;
        }
        tmp->suppress_ifgroup = (int32_t) suppress_ifgroup;
}
```
If you'd like, please add the similar check for `suppress_prefixlen`.

yuwata commented on src/network/networkd-routing-policy-rule.c:
```diff
@@ -1622,6 +1643,54 @@ int config_parse_routing_policy_rule_suppress_prefixlen(
+                return 0;
+        }
+        if (suppress_ifgroup < 0) {
+                log_syntax(unit, LOG_WARNING, filename, line, r,
```
`r` -> `0`

bacher09 replied:
Sorry, I don't understand what have you meant here. `r` should be `0` here, since `safe_atoi32` returns `0` on success, but what I need change here ?

yuwata replied:
Yes, please explicitly set 0 instead of r.

yuwata commented on src/network/networkd-routing-policy-rule.c:
```diff
@@ -1622,6 +1643,54 @@ int config_parse_routing_policy_rule_suppress_prefixlen(
+        }
+        if (suppress_ifgroup < 0) {
+                log_syntax(unit, LOG_WARNING, filename, line, r,
+                           "Value of SuppressIfGroup= is outside of: [0, 2147483647] range: %s", rvalue);
```
"Value of SuppressInterfaceGroup= must be in the range 0…2147483647, ignoring assignment: %s"

yuwata commented on src/network/networkd-setlink.c:
```diff
@@ -458,7 +458,7 @@ static int link_configure(
         }
         case SET_LINK_GROUP:
-                r = sd_netlink_message_append_u32(req, IFLA_GROUP, link->network->group);
+                r = sd_netlink_message_append_u32(req, IFLA_GROUP, (uint32_t)link->network->group);
```
Not a matter, but we usually insert a space after the cast: `(uint32_t) link->network->group`.

yuwata reviewed (changes requested):
Could you rearrange the commits into two? First fix the range of `Network::group`, and then introduce `SuppressInterfaceGroup=`.

bacher09 commented on man/systemd.network.xml:
```diff
@@ -1282,6 +1282,14 @@ Table=1234</programlisting></para>
+          <listitem>
+            <para>Takes a number <replaceable>N</replaceable> in the range of 0…2147483647 and rejects routing
+            decisions that have an interface with the same group id. It has the same meaning as
+            suppress_ifgroup in ip-rule, defaults to unset.</para>
```
Wanted to say that I spelled it as `ip-rule` just because it's the man page that is used by this command.

yuwata replied:
Then, please refer the man page after the sentence.

bacher09 commented:
Thanks @yuwata for the review, I've updated and it should address the most of the comments. Also retested it yet again.

bacher09 commented on src/network/networkd-routing-policy-rule.c:
```diff
@@ -1049,8 +1061,28 @@ int manager_rtnl_process_rule(sd_netlink *rtnl, sd_netlink_message *message, Man
-                tmp->suppress_prefixlen = (int) suppress_prefixlen;
+        if (r >= 0) {
+                // kernel does not limit this value, but we do
+                if (suppress_prefixlen > 128) {
```
I've checked kernel sources and it doesn't limit this value in any way, so in theory possible such situation that some buggy tool will set it over `128`.

yuwata commented on man/systemd.network.xml:
```diff
@@ -1282,6 +1282,14 @@ Table=1234</programlisting></para>
+          <listitem>
+            <para>Takes an integer in the range 0…2147483647 and rejects routing
+            decisions that have an interface with the same group id. It has the same meaning as
+            <option>suppress_ifgroup</option> in <command>ip rule</command>, defaults to unset.</para>
```
`<command>ip rule</command>. Defaults to`

yuwata commented on src/network/networkd-routing-policy-rule.c:
```diff
@@ -1049,8 +1061,28 @@ int manager_rtnl_process_rule(sd_netlink *rtnl, sd_netlink_message *message, Man
-        if (r >= 0)
-                tmp->suppress_prefixlen = (int) suppress_prefixlen;
+        if (r >= 0) {
+                // kernel does not limit this value, but we do
```
Please use `/* */` style.

yuwata reviewed (approved):
Mostly LGTM. Several super minor comments.

Could you add "network: " prefix for each commit title?

yuwata commented:
If possible, please add a testcase in test/test-network/.

yuwata commented:
I force-pushed commits to address the above two points. Setting the green label.

bacher09 commented:
I was already working on a tests (and have fixed these small nits, but haven't pushed it yet). But anyway, I think adding tests in another PR might be good this as well (it basically requires tests for interface `Group`, `SuppressInterfaceGroup` and probably for `SuppressPrefixLength` as well).
